### PR TITLE
Remove PlayMode enum and make Rulesets modular

### PR DESCRIPTION
**Pull request by smoogipoo**
PRing because apparently peppy finally finished this.

### CODE REVIEW

**Comment by smoogipooo** [comment]
NB: I have not reviewed this yet.

File: osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs [hunk 1]
```diff
@@ -135,7 +122,7 @@ private void ensureLoaded(OsuGameBase osu, int timeout = 10000)
                     Thread.Sleep(50);
             };
 
-            Assert.IsTrue(waitAction.BeginInvoke(null, null).AsyncWaitHandle.WaitOne(timeout),
+            Assert.IsTrue(waitAction.BeginInvoke(null, null).AsyncWaitHandle.WaitOne(999999999),
```
Comment (smoogipooo) [inline osu.Game.Tests/Beatmaps/IO/ImportBeatmapTest.cs]: Oops.

**Comment by huoyaoyuan** [comment]
Question: how about referencing another ruleset in beatmap converting?
If tow rulesets provide two-way converting, they must be put into one assembly, for refereceing each other.
Or just don't allow arbitrary converts.

**Comment by peppy** [comment]
@smoogipooo is already working on the conversion code. Basically a ruleset will choose the low-level interface requirements it needs in order to be able to convert (`IHasPosition` for osu! mode, for instance) and if they are all met by at least one object the conversion will be deemed possible.

**Comment by smoogipooo** [comment]
Loading one ruleset loads the others' assembly dependencies.

The final hit object types (e.g. OsuHitObject/TaikoHitObject/etc) should be completely internal to rulesets.
If you wanted Ruleset_B to convert from hit objects of Ruleset_A, Ruleset_A should define an interface, then Ruleset_B references Ruleset_A to use those same interfaces.

As you can imagine, this works best if Ruleset_A defines its interfaces in a "common" assembly, which Ruleset_B then also references, which is actually already being done in the conversion code - osu.Game is the common assembly reference containing those interfaces.

In the end rulesets will require multiple assemblies to be redistributed, including their resources, so we could do some dependency resolution similar to nuget to determine which additional assemblies to redistribute. E.g. if Ruleset_A dependended on OsuRuleset, then we want to make sure the user also downloads OsuRuleset along with it.

**Comment by peppy** [comment]
That kind of conversion shouldn't even be allowed. Interface based or nothing IMO.

**Comment by smoogipooo** [comment]
I agree, however I think that's a topic to be discussed more when we figure out how we're going to perform passes of 3rd party Rulesets efficiently before distributing them to an online database/store.

**Comment by smoogipooo** [comment]
I think this is fine for now, but we need to consider the following cases:

- Consider the case where Ruleset_A is the default ruleset (as per configRuleset). Ruleset_A is then deleted. RulesetDatabase will still return the RulesetInfo for Ruleset_A, and will crash at the point of CreateInstance().  
- Consider the case where all rulesets are deleted.

I think going forward we need to have an "EmptyRuleset" or something, which provides some sane default values and doesn't crash the game.

**Comment by peppy** [comment]
Yep, definitely needs to be more in the direciton of safety when dealing with adding/removing rulesets, but I think it's fine to add that later when it becomes an issue.

**Comment by smoogipooo** [comment]
Steps to reproduce:
1. Delete client.db.
2. Import beatmap.
3. Attempt to enter song select.

Will crash because beatmap.Ruleset is empty in DifficultyIcon.cs. Looks like the foreign key isn't working?

**Review by peppy — APPROVED** [review]
i approve of my own changes!

**Comment by smoogipooo** [comment]
:best:

**Pull request merged**